File: ts/tests/grammar.test.ts
import { describe, it, expect, beforeEach } from "vitest";
import { Grammar } from "../src/grammar.js";
import { InMemoryStore } from "../src/store.js";
import { createBootstrap, NoopSigner } from "../src/event.js";
import { ActorId, ConversationId } from "../src/types.js";

describe("Grammar", () => {
  let store: InMemoryStore;
  let grammar: Grammar;
  const signer = new NoopSigner();
  const alice = new ActorId("alice");
  const bob = new ActorId("bob");
  const convId = new ConversationId("conv_test");

  beforeEach(() => {
    store = new InMemoryStore();
    grammar = new Grammar(store);
    // Bootstrap so there is a chain head
    const boot = createBootstrap(alice, signer);
    store.append(boot);
  });

  // ── emit ──────────────────────────────────────────────────────────

  describe("emit", () => {
    it("creates a grammar.emitted event with the given causes", () => {
      const head = store.head().unwrap();
      const ev = grammar.emit(alice, "hello world", convId, [head.id], signer);
      expect(ev.type.value).toBe("grammar.emitted");
      expect(ev.content.Body).toBe("hello world");
      expect(ev.source.value).toBe("alice");
      expect(store.count()).toBe(2);
    });

    it("throws if causes is empty", () => {
      expect(() => grammar.emit(alice, "hello", convId, [], signer)).toThrow(
        "emit requires at least one cause",
      );
    });

    it("hash-chains to the previous head", () => {
      const headBefore = store.head().unwrap();
      const ev = grammar.emit(alice, "body", convId, [headBefore.id], signer);
      expect(ev.prevHash.value).toBe(headBefore.hash.value);
    });
  });

  // ── respond ───────────────────────────────────────────────────────

  describe("respond", () => {
    it("creates a grammar.responded event with parent in content and causes", () => {
      const head = store.head().unwrap();
      const ev = grammar.respond(alice, "reply", head.id, convId, signer);
      expect(ev.type.value).toBe("grammar.responded");
      expect(ev.content.Body).toBe("reply");
      expect(ev.content.Parent).toBe(head.id.value);
      const causeValues = [...ev.causes].map((c) => c.value);
      expect(causeValues).toContain(head.id.value);
    });
  });

  // ── derive ────────────────────────────────────────────────────────

  describe("derive", () => {
    it("creates a grammar.derived event with Source in content", () => {
      const head = store.head().unwrap();
      const ev = grammar.derive(alice, "derived content", head.id, convId, signer);
      expect(ev.type.value).toBe("grammar.derived");
      expect(ev.content.Body).toBe("derived content");
      expect(ev.content.Source).toBe(head.id.value);
    });
  });

  // ── extend ────────────────────────────────────────────────────────

  describe("extend", () => {
    it("creates a grammar.extended event with Previous in content", () => {
      const head = store.head().unwrap();
      const ev = grammar.extend(alice, "continued", head.id, convId, signer);
      expect(ev.type.value).toBe("grammar.extended");
      expect(ev.content.Body).toBe("continued");
      expect(ev.content.Previous).toBe(head.id.value);
    });
  });

  // ── retract ───────────────────────────────────────────────────────

  describe("retract", () => {
    it("creates a grammar.retracted event when source matches target author", () => {
      const head = store.head().unwrap();
      const emitted = grammar.emit(alice, "to retract", convId, [head.id], signer);
      const ev = grammar.retract(alice, emitted.id, "changed my mind", convId, signer);
      expect(ev.type.value).toBe("grammar.retracted");
      expect(ev.content.Target).toBe(emitted.id.value);
      expect(ev.content.Reason).toBe("changed my mind");
    });

    it("throws if source is not the author of the target event", () => {
      const head = store.head().unwrap();
      const emitted = grammar.emit(alice, "alice's event", convId, [head.id], signer);
      expect(() =>
        grammar.retract(bob, emitted.id, "nope", convId, signer),
      ).toThrow("retract: actor bob cannot retract event");
    });
  });

  // ── annotate ──────────────────────────────────────────────────────

  describe("annotate", () => {
    it("creates a grammar.annotated event with key/value metadata", () => {
      const head = store.head().unwrap();
      const ev = grammar.annotate(alice, head.id, "priority", "high", convId, signer);
      expect(ev.type.value).toBe("grammar.annotated");
      expect(ev.content.Target).toBe(head.id.value);
      expect(ev.content.Key).toBe("priority");
      expect(ev.content.Value).toBe("high");
    });
  });

  // ── merge ─────────────────────────────────────────────────────────

  describe("merge", () => {
    it("creates a grammar.merged event joining multiple sources", () => {
      const head = store.head().unwrap();
      const ev1 = grammar.emit(alice, "branch A", convId, [head.id], signer);
      const ev2 = grammar.emit(alice, "branch B", convId, [ev1.id], signer);
      const merged = grammar.merge(alice, "merged result", [ev1.id, ev2.id], convId, signer);
      expect(merged.type.value).toBe("grammar.merged");
      expect(merged.content.Body).toBe("merged result");
      expect(merged.content.Sources).toEqual([ev1.id.value, ev2.id.value]);
      const causeValues = [...merged.causes].map((c) => c.value);
      expect(causeValues).toContain(ev1.id.value);
      expect(causeValues).toContain(ev2.id.value);
    });

    it("throws if fewer than two sources", () => {
      const head = store.head().unwrap();
      expect(() =>
        grammar.merge(alice, "only one", [head.id], convId, signer),
      ).toThrow("merge requires at least two sources");
    });

    it("throws if zero sources", () => {
      expect(() =>
        grammar.merge(alice, "none", [], convId, signer),
      ).toThrow("merge requires at least two sources");
    });
  });

  // ── chain integrity ───────────────────────────────────────────────

  describe("chain integrity", () => {
    it("maintains valid hash chain across multiple grammar operations", () => {
      const head = store.head().unwrap();
      const ev1 = grammar.emit(alice, "first", convId, [head.id], signer);
      const ev2 = grammar.respond(alice, "second", ev1.id, convId, signer);
      grammar.annotate(alice, ev2.id, "tag", "important", convId, signer);

      const verification = store.verifyChain();
      expect(verification.valid).toBe(true);
      expect(verification.length).toBe(4); // bootstrap + 3 grammar ops
    });
  });
});

File: rust/tests/grammar_test.rs
use eventgraph::event::*;
use eventgraph::grammar::Grammar;
use eventgraph::store::*;
use eventgraph::types::*;

fn setup() -> (InMemoryStore, Event) {
    let mut store = InMemoryStore::new();
    let boot = create_bootstrap(ActorId::new("alice").unwrap(), &NoopSigner, 1);
    let boot = store.append(boot).unwrap();
    (store, boot)
}

// ── emit ────────────────────────────────────────────────────────────────

#[test]
fn emit_creates_emitted_event() {
    let (mut store, boot) = setup();
    let mut g = Grammar::new(&mut store);

    let ev = g
        .emit(
            ActorId::new("alice").unwrap(),
            "hello world",
            ConversationId::new("conv_1").unwrap(),
            vec![boot.id.clone()],
            &NoopSigner,
        )
        .unwrap();

    assert_eq!(ev.event_type.value(), "grammar.emitted");
    assert_eq!(ev.source.value(), "alice");
    let content = ev.content();
    assert_eq!(content["Body"], "hello world");
}

#[test]
fn emit_empty_causes_fails() {
    let (mut store, _boot) = setup();
    let mut g = Grammar::new(&mut store);

    let result = g.emit(
        ActorId::new("alice").unwrap(),
        "hello",
        ConversationId::new("conv_1").unwrap(),
        vec![],
        &NoopSigner,
    );

    assert!(result.is_err());
}

// ── respond ─────────────────────────────────────────────────────────────

#[test]
fn respond_creates_responded_event() {
    let (mut store, boot) = setup();
    let mut g = Grammar::new(&mut store);

    let ev = g
        .respond(
            ActorId::new("bob").unwrap(),
            "I agree",
            boot.id.clone(),
            ConversationId::new("conv_1").unwrap(),
            &NoopSigner,
        )
        .unwrap();

    assert_eq!(ev.event_type.value(), "grammar.responded");
    let content = ev.content();
    assert_eq!(content["Body"], "I agree");
    assert_eq!(content["Parent"], boot.id.value());
}

// ── derive ──────────────────────────────────────────────────────────────

#[test]
fn derive_creates_derived_event() {
    let (mut store, boot) = setup();
    let mut g = Grammar::new(&mut store);

    let ev = g
        .derive(
            ActorId::new("alice").unwrap(),
            "derived insight",
            boot.id.clone(),
            ConversationId::new("conv_1").unwrap(),
            &NoopSigner,
        )
        .unwrap();

    assert_eq!(ev.event_type.value(), "grammar.derived");
    let content = ev.content();
    assert_eq!(content["Body"], "derived insight");
    assert_eq!(content["Source"], boot.id.value());
}

// ── extend ──────────────────────────────────────────────────────────────

#[test]
fn extend_creates_extended_event() {
    let (mut store, boot) = setup();
    let mut g = Grammar::new(&mut store);

    let ev = g
        .extend(
            ActorId::new("alice").unwrap(),
            "more detail",
            boot.id.clone(),
            ConversationId::new("conv_1").unwrap(),
            &NoopSigner,
        )
        .unwrap();

    assert_eq!(ev.event_type.value(), "grammar.extended");
    let content = ev.content();
    assert_eq!(content["Body"], "more detail");
    assert_eq!(content["Previous"], boot.id.value());
}

// ── retract ─────────────────────────────────────────────────────────────

#[test]
fn retract_creates_retracted_event() {
    let (mut store, boot) = setup();
    let mut g = Grammar::new(&mut store);

    // First emit something as alice (bootstrap is also by alice)
    let emitted = g
        .emit(
            ActorId::new("alice").unwrap(),
            "oops",
            ConversationId::new("conv_1").unwrap(),
            vec![boot.id.clone()],
            &NoopSigner,
        )
        .unwrap();

    let ev = g
        .retract(
            ActorId::new("alice").unwrap(),
            emitted.id.clone(),
            "mistake",
            ConversationId::new("conv_1").unwrap(),
            &NoopSigner,
        )
        .unwrap();

    assert_eq!(ev.event_type.value(), "grammar.retracted");
    let content = ev.content();
    assert_eq!(content["Target"], emitted.id.value());
    assert_eq!(content["Reason"], "mistake");
}

#[test]
fn retract_by_non_author_fails() {
    let (mut store, boot) = setup();
    let mut g = Grammar::new(&mut store);

    // boot was authored by alice; bob cannot retract it
    let result = g.retract(
        ActorId::new("bob").unwrap(),
        boot.id.clone(),
        "nope",
        ConversationId::new("conv_1").unwrap(),
        &NoopSigner,
    );

    assert!(result.is_err());
}

// ── annotate ────────────────────────────────────────────────────────────

#[test]
fn annotate_creates_annotated_event() {
    let (mut store, boot) = setup();
    let mut g = Grammar::new(&mut store);

    let ev = g
        .annotate(
            ActorId::new("alice").unwrap(),
            boot.id.clone(),
            "sentiment",
            "positive",
            ConversationId::new("conv_1").unwrap(),
            &NoopSigner,
        )
        .unwrap();

    assert_eq!(ev.event_type.value(), "grammar.annotated");
    let content = ev.content();
    assert_eq!(content["Target"], boot.id.value());
    assert_eq!(content["Key"], "sentiment");
    assert_eq!(content["Value"], "positive");
}

// ── merge ───────────────────────────────────────────────────────────────

#[test]
fn merge_creates_merged_event() {
    let (mut store, boot) = setup();
    let mut g = Grammar::new(&mut store);

    let ev1 = g
        .emit(
            ActorId::new("alice").unwrap(),
            "point one",
            ConversationId::new("conv_1").unwrap(),
            vec![boot.id.clone()],
            &NoopSigner,
        )
        .unwrap();

    let ev2 = g
        .emit(
            ActorId::new("alice").unwrap(),
            "point two",
            ConversationId::new("conv_1").unwrap(),
            vec![ev1.id.clone()],
            &NoopSigner,
        )
        .unwrap();

    let merged = g
        .merge(
            ActorId::new("alice").unwrap(),
            "synthesis",
            vec![ev1.id.clone(), ev2.id.clone()],
            ConversationId::new("conv_1").unwrap(),
            &NoopSigner,
        )
        .unwrap();

    assert_eq!(merged.event_type.value(), "grammar.merged");
    let content = merged.content();
    assert_eq!(content["Body"], "synthesis");
    let sources = content["Sources"].as_array().unwrap();
    assert_eq!(sources.len(), 2);
    assert_eq!(sources[0], ev1.id.value());
    assert_eq!(sources[1], ev2.id.value());
}

#[test]
fn merge_with_fewer_than_two_sources_fails() {
    let (mut store, boot) = setup();
    let mut g = Grammar::new(&mut store);

    let result = g.merge(
        ActorId::new("alice").unwrap(),
        "lone merge",
        vec![boot.id.clone()],
        ConversationId::new("conv_1").unwrap(),
        &NoopSigner,
    );

    assert!(result.is_err());
}

// ── chain integrity ─────────────────────────────────────────────────────

#[test]
fn grammar_operations_maintain_hash_chain() {
    let (mut store, boot) = setup();
    let mut g = Grammar::new(&mut store);

    let e1 = g
        .emit(
            ActorId::new("alice").unwrap(),
            "first",
            ConversationId::new("conv_1").unwrap(),
            vec![boot.id.clone()],
            &NoopSigner,
        )
        .unwrap();

    let e2 = g
        .respond(
            ActorId::new("bob").unwrap(),
            "reply",
            e1.id.clone(),
            ConversationId::new("conv_1").unwrap(),
            &NoopSigner,
        )
        .unwrap();

    let _e3 = g
        .annotate(
            ActorId::new("alice").unwrap(),
            e2.id.clone(),
            "tag",
            "important",
            ConversationId::new("conv_1").unwrap(),
            &NoopSigner,
        )
        .unwrap();

    // Drop grammar to regain access to store
    drop(g);

    let v = store.verify_chain();
    assert!(v.valid);
    assert_eq!(v.length, 4); // boot + 3 grammar events
}
